### DISCUSSION TITLE
Optimizing nested find query [MongoDB]

### DISCUSSION THREAD

**Discussion by bn-piar** [prisma/prisma#16412] [Q&A]
Hey guys. I often use nested querying to find documents by their nested properties. 

The `where` property of a query looks like this
```ts
const where: Prisma.TagWhereInput = {
     // text search
      ...(search
        ? {
            OR: [
              {
                tagId: search
              },
              {
                product: {
                  is: {
                    OR: [// test multiple `product` properties to contain `search`],
                  },
                },
              },
              {
                collection: {
                  is: {
                    collectionName: search,
                  },
                },
              },
              {
                pos: {
                  is: {
                    posName: search,
                  },
                },
              },
            ],
          }
        : {}),
    };
```

the models: 
```prisma
model Tag {
  tagId String @unique
  product Product @relation(...)
  productId String
  collection Collection @relation(...)
  collectionId String
  pos Pos @relation(...)
  posId String
}
model Product {
  barcode String @unique
  productName String
  size String
  tags Tag[]
  collection Collection @relation(...)
  collectionId String
}
model Collection {
  collectionName String @unique
  products Product[]
  tags Tag[]
}
model Pos {
  posName String
  collections Collection[]
  products Product[]
  tags Tag[]
}
```

All properties are indexed but the query takes very long time to complete, 30 seconds +-, so I thought maybe adding the properties I search for (`productBarcode`, `productName`, `collectionName`, etc.) directly on `Tag` model would be a better approach. But then the question arises - how to keep all these properties in sync with the properties on corresponding models. 

Does Prisma provide a tool for keeping properties in sync? Maybe you know of another solution to ridiculously long queries?

ps. I never dealt with databases on a production scale so my models design is probably, well, bad. Please don't throw shit at me

**Comment by pantharshit00** [comment]
Hello @bn-piar 

Your datamodel looks fine to me at a first glance. Can you please try enabling [tracing](https://www.prisma.io/docs/concepts/components/prisma-client/opentelemetry-tracing) and check which queries that Prisma is generating are actually slow? Then we can try to optimise things here.

**Reply by bn-piar** [reply]
couldn't make tracing work. I'll try again this weekend, when I have more spare time

**Reply by vladi-stevanovic** [reply]
:wave: Hello @bn-piar , I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Reply by bn-piar** [reply]
@vladi-stevanovic hi. I'm still facing the issue. Couldn't setup tracing in my app. Is these two queries different performance-wise?
```ts
// 1
findMany({
  where: {
    OR: [
      { product: { productName: { contains: search } } },
      { product: { barcode: { contains: search } } },
      { product: { price: { contains: search } } }
    ]
  }
});
// 2
findMany({
  where: {
    product: {
      is: {
        OR: [
          {  productName: { contains: search } },
          {  barcode: { contains: search } },
          {  price: { contains: search } }
        ]
      }
    ]
  }
});
```

**Reply by nurul3101** [reply]
Are you getting any specific issue while trying to setup tracing? Both the queries intend to achieve the same thing, but to deterministically know which one is better, I would suggest you to enable [logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging) to get the underlying MongoDB query and use [explain](https://www.mongodb.com/docs/manual/reference/method/db.collection.explain/) method to get the query execution plan for both the queries.

**Reply by pantharshit00** [reply]
👋 Hello @bn-piar,  did you perhaps have a chance to check @nurul3101's recommendation? Let us know if you still have any questions!
 
Or, if you've found a solution, we'd love to know how you solved this problem. Feel free to post it in this thread and "Mark as answer" if this discussion is now solved!

**Reply by bn-piar** [reply]
I gave up on setting up logging but while playing around I figured the slow down is caused by `contains` . `equal` operator is like 10-20 times faster than `contains`. We decided to drop the functionality that causes slow down

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
